File: backend/models/Cage.js
const mongoose = require('mongoose');

const CageSchema = new mongoose.Schema({
  cageNumber: {
    type: String,
    required: [true, "Le numéro de la cage est obligatoire"],
    unique: true,
    trim: true
  },
  male: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Bird',
    required: [true, "L'oiseau mâle est obligatoire"]
  },
  female: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Bird',
    required: [true, "L'oiseau femelle est obligatoire"]
  },
  species: {
    type: String,
    required: [true, "L'espèce est obligatoire"],
    trim: true
  },
  status: {
    type: String,
    enum: ['active', 'inactive'],
    default: 'active'
  },
  notes: {
    type: String,
    trim: true
  },
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  }
}, {
  timestamps: true // This will add createdAt and updatedAt fields
});

// Ensure that male and female birds are different
CageSchema.pre('save', async function(next) {
  if (this.male.toString() === this.female.toString()) {
    next(new Error("Le mâle et la femelle doivent être différents"));
  }
  next();
});

module.exports = mongoose.models.Cage || mongoose.model('Cage', CageSchema); 

File: flutter/lib/models/cage.dart
import 'package:app_volailles/models/bird.dart';
import 'package:intl/intl.dart';

class Cage {
  final String? id;
  final Bird? male;
  final Bird? female;
  final String species;
  final DateTime? createdAt;
  final String? status;
  final String? notes;
  final String cageNumber;

  Cage({
    this.id,
    this.male,
    this.female,
    required this.species,
    this.createdAt,
    this.status,
    this.notes,
    required this.cageNumber,
  });

  factory Cage.fromJson(Map<String, dynamic> json) {
    return Cage(
      id: json['_id'],
      male: json['male'] != null ? Bird.fromApi(json['male']) : null,
      female: json['female'] != null ? Bird.fromApi(json['female']) : null,
      species: json['species'] ?? '',
      createdAt:
          json['createdAt'] != null
              ? DateTime.parse(json['createdAt']).toLocal()
              : null,
      status: json['status'] ?? 'active',
      notes: json['notes'],
      cageNumber: json['cageNumber'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'male': male?.toJson(),
      'female': female?.toJson(),
      'species': species,
      'createdAt': createdAt?.toIso8601String(),
      'status': status,
      'notes': notes,
      'cageNumber': cageNumber,
    };
  }

  String get formattedCreatedAt {
    if (createdAt == null) return '';
    final formatter = DateFormat('dd/MM/yyyy HH:mm');
    return formatter.format(createdAt!);
  }
}
